tesseract/tesseract_environment/src/ofkt/ofkt_state_solver.cpp

```diff
@@ -0,0 +1,805 @@
+  }
+
+  if (update_required)
+  {
+    state.link_transforms[node->getLinkName()] = parent_world_tf;
+    state.joint_transforms[node->getJointName()] = parent_world_tf;
+  }
+
+  for (const auto* child : node->getChildren())
+    update(state, child, parent_world_tf, update_required);
+}
+
+void OFKTStateSolver::addNode(const tesseract_scene_graph::Joint::ConstPtr& joint,
+                              const std::string& joint_name,
+                              const std::string& parent_link_name,
+                              const std::string& child_link_name,
+                              std::vector<std::pair<std::string, std::array<double, 2>>>& limits)
+{
+  OFKTNode::UPtr n;
+  switch (joint->type)
+  {
+    case tesseract_scene_graph::JointType::FIXED:
+    {
+      OFKTNode* parent_node = link_map_[parent_link_name];
+      auto n = std::make_unique<OFKTFixedNode>(
+          parent_node, child_link_name, joint_name, joint->parent_to_joint_origin_transform);
+      link_map_[child_link_name] = n.get();
+      parent_node->addChild(n.get());
+      current_state_->link_transforms[n->getLinkName()] = n->getWorldTransformation();
+      current_state_->joint_transforms[n->getJointName()] = n->getWorldTransformation();
+      nodes_[joint_name] = std::move(n);
+
+      break;
+    }
+    case tesseract_scene_graph::JointType::REVOLUTE:
+    {
+      OFKTNode* parent_node = link_map_[parent_link_name];
+      auto n = std::make_unique<OFKTRevoluteNode>(
+          parent_node, child_link_name, joint_name, joint->parent_to_joint_origin_transform, joint->axis);
+      link_map_[child_link_name] = n.get();
+      parent_node->addChild(n.get());
+      current_state_->joints[joint_name] = 0;
+      current_state_->link_transforms[n->getLinkName()] = n->getWorldTransformation();
+      current_state_->joint_transforms[n->getJointName()] = n->getWorldTransformation();
+      nodes_[joint_name] = std::move(n);
+      joint_names_.push_back(joint_name);
+      limits.push_back(
+          std::make_pair(joint_name, std::array<double, 2>({ joint->limits->lower, joint->limits->upper })));
+      break;
+    }
+    case tesseract_scene_graph::JointType::CONTINUOUS:
+    {
+      OFKTNode* parent_node = link_map_[parent_link_name];
+      auto n = std::make_unique<OFKTRevoluteNode>(
+          parent_node, child_link_name, joint_name, joint->parent_to_joint_origin_transform, joint->axis);
+      link_map_[child_link_name] = n.get();
+      parent_node->addChild(n.get());
+      current_state_->joints[joint_name] = 0;
+      current_state_->link_transforms[n->getLinkName()] = n->getWorldTransformation();
+      current_state_->joint_transforms[n->getJointName()] = n->getWorldTransformation();
+      nodes_[joint_name] = std::move(n);
+      joint_names_.push_back(joint_name);
+      limits.push_back(
+          std::make_pair(joint_name, std::array<double, 2>({ joint->limits->lower, joint->limits->upper })));
```

Comment:
Will this work or is some limit assumed? Should a warning be printed?

Reply:
There should always be limits but I have added a check in side the revolute constructor.